### ISSUE TITLE
cannot create object in latest version

### ISSUE THREAD

**Issue by bzuidgeest** [simpleidserver/SimpleIdServer#40]
I just upgraded to 1.1.3 and the post request stopped working with a schema violation error. I gave the application a clean database, so it should not be an "old" data problem.

Schema for eduUser as given by the schema endpoint

{
  "id": "urn:ietf:params:scim:schemas:extension:nleducation:1.0:eduuser",
  "name": "eduUser",
  "description": "SCIM extended resource for representing users for digital educational content in the Netherlands",
  "attributes": [
    {
      "name": "name",
      "type": "complex",
      "multiValued": false,
      "description": "A complex type containing the name of the EduUser",
      "required": true,
      "caseExact": false,
      "mutability": "readwrite",
      "returned": "default",
      "uniqueness": "none",
      "subAttributes": [
        {
          "name": "familyName",
          "type": "string",
          "multiValued": false,
          "description": "The surname of the EduUser, including prefixes like 'van' or 'van der'",
          "required": true,
          "caseExact": false,
          "mutability": "readwrite",
          "returned": "always",
          "uniqueness": "none"
        },
        {
          "name": "givenName",
          "type": "string",
          "multiValued": false,
          "description": "The given name of the EduUser",
          "required": true,
          "caseExact": false,
          "mutability": "readwrite",
          "returned": "always",
          "uniqueness": "none"
        }
      ]
    },
    {
      "name": "eckId",
      "type": "string",
      "multiValued": false,
      "description": "The unique identification of the user for digital content",
      "required": true,
      "caseExact": false,
      "mutability": "immutable",
      "returned": "request",
      "uniqueness": "global"
    },
    {
      "name": "familyName",
      "type": "string",
      "multiValued": false,
      "description": "The surname of the EduUser, including prefixes like 'van' or 'van der'",
      "required": true,
      "caseExact": false,
      "mutability": "readwrite",
      "returned": "always",
      "uniqueness": "none"
    },
    {
      "name": "givenName",
      "type": "string",
      "multiValued": false,
      "description": "The given name of the EduUser",
      "required": true,
      "caseExact": false,
      "mutability": "readwrite",
      "returned": "always",
      "uniqueness": "none"
    },
    {
      "name": "externalId",
      "type": "string",
      "multiValued": false,
      "description": "The identifaction of the user as defined by Kennisnet: nlEduPersonRealId",
      "required": false,
      "caseExact": true,
      "mutability": "readwrite",
      "returned": "always",
      "uniqueness": "global"
    }
  ]
}

The post body to eduUser endpoint
{
    "schemas": [
        "urn:ietf:params:scim:schemas:extension:nleducation:1.0:eduuser"
    ],
    "eckId": "https://eckid.nl/8e0a9f57fc7e7cc810f03faX",
    "externalId": "123ddfs457@magis.main",
    "name": {
        "givenName": "Voornaam",
        "familyName": "Achternaam"
    }
}

The result :
{
  "status": "400",
  "response": {
    "schemas": [
      "urn:ietf:params:scim:api:messages:2.0:Error"
    ],
    "status": "400",
    "scimType": "schemaViolated",
    "detail": "required attributes familyName,givenName are missing"
  }
}

It still worked in 1.1.2. Am I missing something obvious or is there a bug in the 1.1.3 version?
I'm using the ef persistence option package.

**Comment by simpleidserver** [comment]
An exception is thrown because the parameters "givenName" and "familyName" are defined in two places : 
=> root object & "user" complex object.

If you remove both parameters "givenName" and "familyName" from the root object then your request will work:

 ```
{
  "id": "urn:ietf:params:scim:schemas:extension:nleducation:1.0:eduuser",
  "name": "eduUser",
  "description": "SCIM extended resource for representing users for digital educational content in the Netherlands",
  "attributes": [
    {
      "name": "name",
      "type": "complex",
      "multiValued": false,
      "description": "A complex type containing the name of the EduUser",
      "required": true,
      "caseExact": false,
      "mutability": "readwrite",
      "returned": "default",
      "uniqueness": "none",
      "subAttributes": [
        {
          "name": "familyName",
          "type": "string",
          "multiValued": false,
          "description": "The surname of the EduUser, including prefixes like 'van' or 'van der'",
          "required": true,
          "caseExact": false,
          "mutability": "readwrite",
          "returned": "always",
          "uniqueness": "none"
        },
        {
          "name": "givenName",
          "type": "string",
          "multiValued": false,
          "description": "The given name of the EduUser",
          "required": true,
          "caseExact": false,
          "mutability": "readwrite",
          "returned": "always",
          "uniqueness": "none"
        }
      ]
    },
    {
      "name": "eckId",
      "type": "string",
      "multiValued": false,
      "description": "The unique identification of the user for digital content",
      "required": true,
      "caseExact": false,
      "mutability": "immutable",
      "returned": "request",
      "uniqueness": "global"
    },
    {
      "name": "externalId",
      "type": "string",
      "multiValued": false,
      "description": "The identifaction of the user as defined by Kennisnet: nlEduPersonRealId",
      "required": false,
      "caseExact": true,
      "mutability": "readwrite",
      "returned": "always",
      "uniqueness": "global"
    }
  ]
}
```

**Comment by bzuidgeest** [comment]
You are right those properties are duplicated in the schema (I did not notice), but when upgrading from 1.1.2 to 1.1.3 the schema definitions did not change. 

SELECT * FROM [SCIM].[dbo].[SCIMSchemaLst]
gives :
Id	Name	Description	IsRootSchema	ResourceType
urn:ietf:params:scim:schemas:core:2.0:Group	Group	Group	1	Groups
urn:ietf:params:scim:schemas:core:2.0:User	User	User Account	1	Users
urn:ietf:params:scim:schemas:extension:nleducation:1.0:eduuser	eduUser	SCIM extended resource for representing users for digital educational content in the Netherlands	1	eduUsers

And 

select * from SCIMSchemaAttributeModel WHERE SchemaId = 'urn:ietf:params:scim:schemas:extension:nleducation:1.0:eduuser'
gives
Id	ParentId	SchemaId	Name	Type
45c76e9a-fc99-4257-a289-1a644a864c1a	NULL	urn:ietf:params:scim:schemas:extension:nleducation:1.0:eduuser	name	5
4bd40e0f-5610-4cc6-b82f-f5e29a628ef2	NULL	urn:ietf:params:scim:schemas:extension:nleducation:1.0:eduuser	eckId	0
52103a09-c886-4d1d-96c1-45f38590b4ec	45c76e9a-fc99-4257-a289-1a644a864c1a	urn:ietf:params:scim:schemas:extension:nleducation:1.0:eduuser	familyName	0
db97a380-70d9-4eaf-bfbe-80e826386e1f	45c76e9a-fc99-4257-a289-1a644a864c1a	urn:ietf:params:scim:schemas:extension:nleducation:1.0:eduuser	givenName	0
e06792fb-6c61-4c61-a049-993e121d7d48	NULL	urn:ietf:params:scim:schemas:extension:nleducation:1.0:eduuser	externalId	0

sorry about the bad table formatting.
But as you can see familyName is returned only once for this object and it a child of Name. So I submit the library as a error in processing the schema somewhere? Maybe for some reason all subproperties are also added to the root.

**Comment by bzuidgeest** [comment]
I found the underlying cause I think. In 1.1.3 SCIMSchemaAttributeModel  has the schemaId set for every attribute, both parent and child of a object. while 1.1.2 only sets schemaId for the root objects. So the functions creating the schema in the db must be functionaly out of sync with the functions retrieving them.  Thus the query above would only return 3 rows in 1.1.2 and 5 in 1.1.3

**Comment by simpleidserver** [comment]
Correct ! there is an issue in the SQLServer repository, instead of returning 3 attributes, 5 attributes are returned.
We are going to fix the issue ASAP.

**Comment by simpleidserver** [comment]
Can-you check if with the pre-release version "1.1.4-ci-00129" you still have the error ?

https://www.myget.org/feed/advance-ict/package/nuget/SimpleIdServer.Scim.Persistence.EF
https://www.myget.org/feed/advance-ict/package/nuget/SimpleIdServer.Scim

**Comment by bzuidgeest** [comment]
Well compliments for the fast response. I have tried these packages and they seem to solve the problem. My "test" steps run again and I only have a unrelated problem with .search for which I will open another issue ( to keep things clean)
